pkg/dockerregistry/server/blobdescriptorservice.go

```diff
@@ -25,6 +38,140 @@ type blobDescriptorService struct {
 	distribution.BlobDescriptorService
 }
 
+// Stat returns a a blob descriptor if the given blob is either linked in repository or is referenced in
+// corresponding image stream. This method is invoked from inside of upstream's linkedBlobStore. It expects
+// a proper repository object to be set on given context by upper openshift middleware wrappers.
 func (bs *blobDescriptorService) Stat(ctx context.Context, dgst digest.Digest) (distribution.Descriptor, error) {
-	return dockerRegistry.BlobStatter().Stat(ctx, dgst)
+	// if there is a repo layer link, return its desc
+	desc, err := bs.BlobDescriptorService.Stat(ctx, dgst)
+	if err == nil {
+		return desc, err
+	}
+
+	// verify the blob is stored locally
+	desc, err = dockerRegistry.BlobStatter().Stat(ctx, dgst)
+	if err != nil {
+		return desc, err
+	}
+
+	// verify that an image holding this blob is actually present in the image stream
+	repo, found := RepositoryFrom(ctx)
+	if !found {
+		context.GetLogger(ctx).Errorf("failed to retrieve repository from context")
+		return distribution.Descriptor{}, distribution.ErrBlobUnknown
+	}
+
+	if imageStreamHasBlob(repo, dgst) {
+		return desc, err
+	}
+
+	return distribution.Descriptor{}, distribution.ErrBlobUnknown
+}
+
+// imageStreamHasBlob returns true if the given blob digest is referenced in image stream corresponding to
+// given repository. If not found locally, image stream's images will be iterated and fetched from newest to
+// oldest until found. Each processed image will update local cache of blobs.
+func imageStreamHasBlob(r *repository, dgst digest.Digest) bool {
+	repositories := r.cachedLayers.RepositoriesForDigest(dgst)
+	match := localRepositoryReferenceName(r)
+	for _, repo := range repositories {
+		if repo == match {
+			context.GetLogger(r.ctx).Debugf("found cached blob %q in repository %s/%s", dgst.String(), r.namespace, r.name)
+			return true
+		}
+	}
+
+	// verify directly with etcd
+	is, err := r.getImageStream()
+	if err != nil {
+		context.GetLogger(r.ctx).Errorf("failed to get image stream: %v", err)
+		return false
+	}
+
+	tagEvents := []*imageapi.TagEvent{}
+	event2Name := make(map[*imageapi.TagEvent]string)
+	for name, eventList := range is.Status.Tags {
+		for i := range eventList.Items {
+			event := &eventList.Items[i]
+			tagEvents = append(tagEvents, event)
+			event2Name[event] = name
+		}
+	}
+	// search from youngest to oldest
+	sort.Sort(ByGeneration(tagEvents))
+
+	processedBlobs := map[string]struct{}{}
+	processedImages := map[string]struct{}{}
+
+	for _, tagEvent := range tagEvents {
+		if _, processed := processedImages[tagEvent.Image]; processed {
+			continue
+		}
+		if imageHasBlob(r, processedBlobs, match, tagEvent.Image, dgst.String(), !r.pullthrough) {
+			tagName := event2Name[tagEvent]
```

Comment:
> but we might only want to cache the image that we actually found's layers. Is there a case where a user is going to pull two images from the same image stream in close succession?

Good catch. It seemed to me like a waste to throw away the fetched data. But using this approach we may push out frequently accessed items if the cache is close to full.

> Also, you added rememberLayersOfImage below but you don't use it from imageHasBlob?

I planned to use it but it proved as not useful in the end. I'll remove it.